### ISSUE TITLE
timeout [regression] Exception in multiple nxos_* modules when using SSH transport

### ISSUE THREAD

**Issue by jean-christophe-manciot** [ansible/ansible#18974]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

nxos_nxapi
nxos_command
nxos_template
##### ANSIBLE VERSION

```
ansible --version
ansible 2.3.0
```
##### CONFIGURATION

inventory = ./hosts
roles_path = /home/actionmystique/Program-Files/Ubuntu/Ansible/git-Ansible/Roles/roles
fact_caching = redis
fact_caching_timeout = 86400
##### OS / ENVIRONMENT
- **Local host**: Ubuntu 16.10 4.8
- **Target nodes**: 
  - NX-OSv 7.3(0)D1(1) (last release available in Cisco VIRL)
  - NX-OSv 7.2(0)
##### SUMMARY

Running nxos_nxapi triggers an exception, whereas I can manually login into the target node with SSH.
##### STEPS TO REPRODUCE

**Inventory ./hosts**:

```
[nx_osv]
192.168.137.244 ansible_host=NX_OSv_1 mgt_interface=mgmt0
192.168.137.234 ansible_host=NX_OSv_2 mgt_interface=mgmt0
```

Structure passed as "**provider**": connections.ssh
Defined in group_vars/nx_osv/connections.yml

```
connections:
        compute: local
...
        ssh:
          transport: cli 
          host: "{{ inventory_hostname }}"
          # ansible_port
          port: 22
          # ansible_user
          username: admin
          # ansible_ssh_pass
          password: xxxxxxxxxxx
          authorize: yes
          # enable_secret_password
          auth_pass: xxxxxxxxxx
          # private_key_file
          ssh_keyfile: "~/.ssh/id_rsa"
          version: 2
          timeout: 10
ansible_connection: "{{ connections.compute }}" 
```

Structures http & https
Defined in group_vars/nx_osv/http.yml

```
http:
        # enable_http can be true / false
        service: true
        port: 80
https:
        # enable_https can be true / false
        service: false
        port: 10443
```

**Role**: nxos_init

```
- include_vars: "{{ role_path }}/../../group_vars/{{ hostvars[inventory_hostname].group_names[0] }}/connections.yml"
- include_vars: "{{ role_path }}/../../group_vars/{{ hostvars[inventory_hostname].group_names[0] }}/http.yml"

- name: Enable NX-API
  nxos_nxapi: 
      state=started 
      sandbox=false
      http={{ http.service }}
      http_port={{ http.port }}
      https={{ https.service }}
      https_port={{ https.port }}
      provider={{ connections.ssh }}
  register: return
...
```

**Playbook**: init.yml

```
- name: Initialize all NX-OS features & save facts
  hosts:
    - nx_os
    - nx_osv
  roles:
    - nxos_init
```
##### EXPECTED RESULTS

Successful nxos_nxapi
##### ACTUAL RESULTS

```
TASK [nxos_init : Enable NX-API] ***********************************************
task path: /home/actionmystique/Program-Files/Ubuntu/Ansible/git-Ansible/roles/nxos_init/tasks/main.yml:29
<NX_OSv_1> ESTABLISH LOCAL CONNECTION FOR USER: root
<NX_OSv_1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1470230351.99-196486228076050 `" && echo ansible-tmp-1470230351.99-196486228076050="` echo $HOME/.ansible/tmp/ansible-tmp-1470230351.99-196486228076050 `" ) && sleep 0'
<NX_OSv_2> ESTABLISH LOCAL CONNECTION FOR USER: root
<NX_OSv_2> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1470230351.99-268650786865216 `" && echo ansible-tmp-1470230351.99-268650786865216="` echo $HOME/.ansible/tmp/ansible-tmp-1470230351.99-268650786865216 `" ) && sleep 0'
<NX_OSv_2> PUT /tmp/tmpu2bixy TO /root/.ansible/tmp/ansible-tmp-1470230351.99-268650786865216/nxos_nxapi
<NX_OSv_1> PUT /tmp/tmpyyn3Hd TO /root/.ansible/tmp/ansible-tmp-1470230351.99-196486228076050/nxos_nxapi
<NX_OSv_2> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1470230351.99-268650786865216/nxos_nxapi; rm -rf "/root/.ansible/tmp/ansible-tmp-1470230351.99-268650786865216/" > /dev/null 2>&1 && sleep 0'
<NX_OSv_1> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1470230351.99-196486228076050/nxos_nxapi; rm -rf "/root/.ansible/tmp/ansible-tmp-1470230351.99-196486228076050/" > /dev/null 2>&1 && sleep 0'
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_4CyDsX/ansible_module_nxos_nxapi.py", line 257, in <module>
    main()
  File "/tmp/ansible_4CyDsX/ansible_module_nxos_nxapi.py", line 248, in main
    result = config_server(module)
  File "/tmp/ansible_4CyDsX/ansible_module_nxos_nxapi.py", line 156, in config_server
    nxapi_state = get_nxapi_state(module)
  File "/tmp/ansible_4CyDsX/ansible_module_nxos_nxapi.py", line 148, in get_nxapi_state
    features = module.execute(['show feature | grep nxapi'])[0]
  File "/tmp/ansible_4CyDsX/ansible_modlib.zip/ansible/module_utils/nxos.py", line 253, in execute
  File "/tmp/ansible_4CyDsX/ansible_modlib.zip/ansible/module_utils/nxos.py", line 230, in connect
  File "/tmp/ansible_4CyDsX/ansible_modlib.zip/ansible/module_utils/nxos.py", line 182, in connect
  File "/tmp/ansible_4CyDsX/ansible_modlib.zip/ansible/module_utils/shell.py", line 120, in open
  File "/tmp/ansible_4CyDsX/ansible_modlib.zip/ansible/module_utils/shell.py", line 129, in receive
  File "/usr/local/lib/python2.7/dist-packages/paramiko/channel.py", line 615, in recv
    raise socket.timeout()
socket.timeout
...
```
##### Troubleshooting
- Downgrading paramiko from 2.0.1 to 1.16 does not change anything
- When trying to manually ssh into the target:

```
# ssh admin@192.168.137.234
User Access Verification
Cisco NX-OS Software
Copyright (c) 2002-2016, Cisco Systems, Inc. All rights reserved.
NX-OSv software ("NX-OSv Software") and related documentation,
...
NX_OSv_2# exit
Connection to 192.168.137.234 closed.
```

**Comment by jean-christophe-manciot** [comment]
This issue has been originally posted in ansible//ansible-modules-core [here](https://github.com/ansible/ansible-modules-core/issues/4311#issuecomment-265521860).

**Comment by ansibot** [comment]
cc @jedelman8, @GGabriele, @privateip, @gundalow, @qalthos
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by privateip** [comment]
Based on the output this bug occurred before we implemented the new network_cli conenction pluign.  I  successfully tested timeout values with both cli and nxapi transports in the latest devel branch and could make tasks either pass or fail based on the time.  Could you re-test with the latest devel and report back either working (and close the issue) or update the status?

**Comment by jean-christophe-manciot** [comment]
Nope, the issue is still there, even with ```timeout: 60``` in the ```provider``` structure (not sure if it's implemented or not, the [documentation](http://docs.ansible.com/ansible/nxos_nxapi_module.html) does not say) and:
```
- name: Enable NX-API
  nxos_nxapi: 
      state: present 
      sandbox: "{{ sandbox }}"
      http: "{{ http.service }}"
      http_port: "{{ http.port }}"
      https: "{{ https.service }}"
      https_port: "{{ https.port }}"
      provider: "{{ connections.ssh }}"
  register: nxapi
  ignore_errors: yes
```
leads to:
```
TASK [nxos_init : Enable NX-API] ********************************************************************************************************************************************
task path: /home/actionmystique/Program-Files/Ubuntu/Ansible/Roles/roles/nxos_init/tasks/main.yml:70
fatal: [NX_OSv_Spine_12]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell", 
    "rc": 255
}
```
I can ssh into that target.

**Comment by privateip** [comment]
Ok could you try something please?   Enabling ansible logging and run your playbook with `-vvvv` (4) option.  To enable logging do the following in terminal:

```
export ANSIBLE_LOG_PATH=$PWD/ansible.log
```

This will create a log file in the director you are in called `ansible.log`

Then run your playbook as such `ansible-playbook -vvvv <pb.yaml>`

Finally grep the log file with `grep ansible.log -e "^<"` and add that to this issue?

Thanks

**Comment by privateip** [comment]
@jean-christophe-manciot are we good to close this issue?

**Comment by jean-christophe-manciot** [comment]
````
<172.21.100.12> starting new persistent socket with path /root/.ansible/pc/a27a040b59
<172.21.100.12> using connection plugin network_cli
<172.21.100.12> ESTABLISH CONNECTION FOR USER: admin on PORT 22 TO 172.21.100.12
<172.21.100.12> ssh connection done, setting terminal
<172.21.100.12> ssh connection has completed successfully
<172.21.100.12> incoming request accepted on persistent socket
<172.21.100.12> received data on socket with len 1562
<172.21.100.12> socket operation is CONTEXT
<172.21.100.12> updating play_context for connection
<172.21.100.12> received data on socket with len 18
<172.21.100.12> socket operation is EXEC
<172.21.100.12> attempting to open shell to device
<172.21.100.12> closing shell due to sigalarm
<172.21.100.12> closing shell on device
<172.21.100.12> socket operation completed with rc 255
<172.21.100.12> shutting down control socket, connection was active for 0:00:11.001872 secs
<172.21.100.12> closing connection
<172.21.100.12> closing shell on device
```

**Comment by jean-christophe-manciot** [comment]
Still there (same symptom) with ansible **2.3.0 installed today** with ```pip install git+git://github.com/ansible/ansible.git@devel```.

**Comment by jean-christophe-manciot** [comment]
With ansible **2.2.1.0**, the symptom is different:
```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_Bnc2GZ/ansible_module_nxos_nxapi.py", line 326, in <module>
    main()
  File "/tmp/ansible_Bnc2GZ/ansible_module_nxos_nxapi.py", line 310, in main
    instance = get_instance(module)
  File "/tmp/ansible_Bnc2GZ/ansible_module_nxos_nxapi.py", line 157, in get_instance
    resp = module.cli('show nxapi', 'json')
  File "/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/network.py", line 117, in cli
  File "/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/network.py", line 148, in connect
  File "/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/nxos.py", line 266, in connect
  File "/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/shell.py", line 226, in connect
  File "/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/shell.py", line 116, in open
  File "/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/shell.py", line 128, in receive
  File "/usr/local/lib/python2.7/dist-packages/paramiko/channel.py", line 667, in recv
    raise socket.timeout()
socket.timeout

fatal: [NX_OSv_Spine_12]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_name": "nxos_nxapi"
    }, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_Bnc2GZ/ansible_module_nxos_nxapi.py\", line 326, in <module>\n    main()\n  File \"/tmp/ansible_Bnc2GZ/ansible_module_nxos_nxapi.py\", line 310, in main\n    instance = get_instance(module)\n  File \"/tmp/ansible_Bnc2GZ/ansible_module_nxos_nxapi.py\", line 157, in get_instance\n    resp = module.cli('show nxapi', 'json')\n  File \"/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/network.py\", line 117, in cli\n  File \"/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/network.py\", line 148, in connect\n  File \"/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/nxos.py\", line 266, in connect\n  File \"/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/shell.py\", line 226, in connect\n  File \"/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/shell.py\", line 116, in open\n  File \"/tmp/ansible_Bnc2GZ/ansible_modlib.zip/ansible/module_utils/shell.py\", line 128, in receive\n  File \"/usr/local/lib/python2.7/dist-packages/paramiko/channel.py\", line 667, in recv\n    raise socket.timeout()\nsocket.timeout\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE"
}
```

**Comment by ansibot** [comment]
cc @rcarrillocruz
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by beggles** [comment]
I too am having trouble running plays to NXOS devices (NX-OSv9K v7.0(3)I5(2)) since upgrading to Ansible 2.3, unfortunately I am unsure of which 2.2 build used to work. I am getting the following errors when running a simple "nxos_command" playbook, as follows;

```
---
- hosts: 192.168.0.222
  connection: local

  tasks:
    - name: show ios command
      nxos_command:
        commands:
          - show vlan
          - sh ip int brief
          - sh int status
          - sh int desc
```

with error output as follows;

```
<192.168.0.222> connection transport is cli
<192.168.0.222> socket_path: /home/jason/.ansible/pc/9eee8ed137
open_shell() returned 255  Traceback (most recent call last):
  File "/usr/bin/ansible-connection", line 198, in run
    (rc, stdout, stderr) = self.conn.exec_command(cmd)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/connection/network_cli.py", line 251, in exec_command
    return self.open_shell()
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/connection/__init__.py", line 54, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/connection/network_cli.py", line 106, in open_shell
    self.receive()
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/connection/network_cli.py", line 144, in receive
    data = self._shell.recv(256)
AttributeError: 'NoneType' object has no attribute 'recv'

```
These same error messages have also been recorded in "ansible.log", when running a "nxos_facts" play.

output of `ansible --version` is as follows;

```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

Finally, and for what it is worth, similar plays against an old IOS switch work fine.

**Comment by ansibot** [comment]
cc @rahushen
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jean-christophe-manciot** [comment]
Issue still unsolved with:
- latest stable **ansible 2.3.0**
- latest unstable **ansible 2.4.0 from pip**
```
fatal: [NX_OSv_Spine_31]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell", 
    "rc": 255
}
```
although the target node is accessible through ssh and some other plays run successfully.

**Comment by beggles** [comment]
I updated my playbook to include `transport: nxapi,` and that worked, though when I included an additional task to execute `show vlan` via `transport:cli`, I received the "unable to open shell" error;

New playbook looks like this;
```
---
- hosts: 192.168.0.222
  connection: local

  tasks:
    - name: show nxos command via nxapi
      nxos_command:
        transport: nxapi
        commands:
          - show vlan
          - sh ip int brief
          - sh int status
          - sh int desc

    - name: show nxos command via cli
      nxos_command:
        transport: cli
        commands:
          - show vlan
```

The second task "show nxos command via cli", is the failing task;
```
fatal: [192.168.0.222]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell", 
    "rc": 255
}
```

First task executed fine and as expected.

**Comment by ansibot** [comment]
cc @mikewiebe
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gundalow** [comment]
@beggles Please follow the steps detailed in https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell and raise a separate issue if needed.

**Comment by beggles** [comment]
Thanks @gundalow , please refer to issue 23881, raised last weekend, output captured there and fix confirmed by @trishnaguha, due for release next week as part of 2.3.1 
https://github.com/ansible/ansible/issues/23881

**Comment by trishnaguha** [comment]
resolved_by_pr #23589

**Comment by jean-christophe-manciot** [comment]
@gundalow @trishnaguha 
It is NOT solved with latest unstable ansible from **pip** (pip install git+git://github.com/ansible/ansible.git@devel):
With **NX-OSv 7.3(0)D1(1)** and **NX-OSv-9k 7.0(3)I5(2)**:
```
fatal: [NX_OSv_Spine_11]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell", 
    "rc": 255
}
```
despite being able to manually ssh into both devices:
```
# ssh admin@172.21.100.31
User Access Verification
...
NX_OSv_Spine_31# 
```

With latest **deb build** from latest commit 08a3ea22210f17731ea94c077625cc9bb85dd45b on devel, I get the same issue with the same symptoms.

Have you really tested your patch before claiming that the issue is solved?

**Comment by gundalow** [comment]
@jean-christophe-manciot Could you please post the full error from `devel` using the process here https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell

`unable to open shell` is a "catch all" for a number of possible issues.

needs_info

**Comment by jean-christophe-manciot** [comment]
@gundalow 
The log shows ```closing shell due to sigalarm``` & ```AttributeError: 'NoneType' object has no attribute 'recv'```:
```
2017-05-09 15:43:57,334 p=22961 u=root |  TASK [nxos_init : Enable NX-API] ********************************************************************************************************************************************
2017-05-09 15:43:57,334 p=22961 u=root |  task path: /home/actionmystique/Ansible/Roles/roles/nxos_init/tasks/main.yml:71
2017-05-09 15:43:58,318 p=23064 u=root |  starting new persistent socket with path /root/.ansible/pc/314588aa1e
2017-05-09 15:43:58,318 p=23064 u=root |  current working directory is /media/actionmystique/SAMSUNG-850-Ext4/Labs/GNS3/CCNP-Cloud/CLDAUT/Cisco-Nexus-9k/7.0.3.15.2.free/Ansible
2017-05-09 15:43:58,318 p=23064 u=root |  using connection plugin network_cli
2017-05-09 15:43:58,388 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,388 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,388 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,388 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,388 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,734 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,735 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,735 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,735 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,735 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,735 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,735 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,735 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,740 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,742 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-05-09 15:43:58,749 paramiko.transport starting thread (client mode): 0x56ff2050L
2017-05-09 15:43:58,749 paramiko.transport Local version/idstring: SSH-2.0-paramiko_2.1.2
2017-05-09 15:43:58,795 paramiko.transport Remote version/idstring: SSH-2.0-OpenSSH_6.2 PKIX FIPS
2017-05-09 15:43:58,795 paramiko.transport Connected (version 2.0, client OpenSSH_6.2)
2017-05-09 15:43:58,796 paramiko.transport kex algos:[u'ecdh-sha2-nistp256', u'ecdh-sha2-nistp384', u'ecdh-sha2-nistp521', u'diffie-hellman-group-exchange-sha256', u'diffie-hellman-group-exchange-sha1', u'diffie-hellman-group14-sha1', u'diffie-hellman-group1-sha1'] server key:[u'ssh-rsa'] client encrypt:[u'aes128-ctr', u'aes192-ctr', u'aes256-ctr'] server encrypt:[u'aes128-ctr', u'aes192-ctr', u'aes256-ctr'] client mac:[u'hmac-sha1'] server mac:[u'hmac-sha1'] client compress:[u'none', u'zlib@openssh.com'] server compress:[u'none', u'zlib@openssh.com'] client lang:[u''] server lang:[u''] kex follows?False
2017-05-09 15:43:58,796 paramiko.transport Kex agreed: diffie-hellman-group1-sha1
2017-05-09 15:43:58,796 paramiko.transport Cipher agreed: aes128-ctr
2017-05-09 15:43:58,796 paramiko.transport MAC agreed: hmac-sha1
2017-05-09 15:43:58,796 paramiko.transport Compression agreed: none
2017-05-09 15:43:58,828 paramiko.transport kex engine KexGroup1 specified hash_algo <built-in function openssl_sha1>
2017-05-09 15:43:58,829 paramiko.transport Switch to new keys ...
2017-05-09 15:43:58,934 paramiko.transport userauth is OK
2017-05-09 15:43:58,949 paramiko.transport Auth banner: User Access Verification

2017-05-09 15:43:58,949 paramiko.transport Authentication (password) successful!
2017-05-09 15:43:58,952 p=23064 u=root |  ssh connection done, setting terminal
2017-05-09 15:43:58,956 p=23064 u=root |  ssh connection has completed successfully
2017-05-09 15:43:58,956 p=23064 u=root |  connection established in 0:00:00.637215
2017-05-09 15:43:59,318 p=23064 u=root |  incoming request accepted on persistent socket
2017-05-09 15:43:59,318 p=23064 u=root |  socket operation is CONTEXT
2017-05-09 15:43:59,318 p=23064 u=root |  updating play_context for connection
2017-05-09 15:43:59,319 p=23064 u=root |  socket operation is EXEC
2017-05-09 15:43:59,319 p=23064 u=root |  attempting to open shell to device
2017-05-09 15:43:59,319 paramiko.transport [chan 0] Max packet in: 32768 bytes
2017-05-09 15:43:59,320 paramiko.transport [chan 0] Max packet out: 32768 bytes
2017-05-09 15:43:59,320 paramiko.transport Secsh channel 0 opened.
2017-05-09 15:43:59,323 paramiko.transport [chan 0] Sesch channel 0 request ok
2017-05-09 15:43:59,325 paramiko.transport [chan 0] Sesch channel 0 request ok
2017-05-09 15:44:09,319 p=23064 u=root |  **closing shell due to sigalarm**
2017-05-09 15:44:09,319 p=23064 u=root |  closing shell on device
2017-05-09 15:44:09,319 paramiko.transport [chan 0] EOF sent (0)
2017-05-09 15:44:09,321 p=23064 u=root |  socket operation completed with rc 255
2017-05-09 15:44:09,322 p=23064 u=root |  shutting down control socket, connection was active for 0:00:11.003136 secs
2017-05-09 15:44:09,322 p=23064 u=root |  closing connection
2017-05-09 15:44:09,322 p=23064 u=root |  closing shell on device
2017-05-09 15:44:09,332 paramiko.transport EOF in transport thread
2017-05-09 15:44:09,345 p=22961 u=root |  open_shell() returned 255  Traceback (most recent call last):
  File "/usr/local/bin/ansible-connection", line 192, in run
    (rc, stdout, stderr) = self.conn.exec_command(cmd)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/connection/network_cli.py", line 250, in exec_command
    return self.open_shell()
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/connection/__init__.py", line 52, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/connection/network_cli.py", line 105, in open_shell
    self.receive()
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/connection/network_cli.py", line 143, in receive
    data = self._shell.recv(256)
**AttributeError: 'NoneType' object has no attribute 'recv'**

2017-05-09 15:44:09,348 p=22961 u=root |  fatal: [NX_OSv_Spine_31]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell", 
    "rc": 255
}
```

**Comment by ganeshrn** [comment]
@jean-christophe-manciot  Based on error message `"msg": "unable to open shell",` i suspect you are running with old code base.
The error message should include an `url`  as below.

`"msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"`

It was added in this commit https://github.com/ansible/ansible/commit/529df8640ba80eded2ed9671705f3e487f2eee6a

**Comment by jean-christophe-manciot** [comment]
@ganeshrn
The previous lines were logged with ansible ```2.3.0.0-1``` built from sources.
I get the exact same log with latest ansible ```2.4.0``` from "pip install git+git://github.com/ansible/ansible.git@devel". 

How can we check which latest commit hash corresponds to the pip version?

**Comment by ganeshrn** [comment]
Easier way to work with Github code is
```
git clone https://github.com/ansible/ansible
cd ansible
source hacking/env-setup
```

After executing above commands output of
```
ansible --version
```
should point to latest commit on devel branch.

And output of
```
which ansible
```
should point to a path in cloned repository.

To work with `2.3` you can checkout `stable-2.3` branch

**Comment by jean-christophe-manciot** [comment]
You have not answered my question, which makes me suspect that the **pip builds may not be maintained anymore, or incorrectly at least**, because I explicitly run the main script from /usr/local/bin and:
```
/usr/local/bin/ansible --version
ansible 2.4.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/actionmystique/Ansible/git-yang-networkop/ansible-101/library']
  python version = 2.7.12+ (default, Sep 17 2016, 12:08:02) [GCC 6.2.0 20160914]
```
and
```
# pip show ansible
Name: ansible
Version: 2.4.0
Summary: Radically simple IT automation
Home-page: https://ansible.com/
Author: Ansible, Inc.
Author-email: info@ansible.com
License: GPLv3
Location: /usr/local/lib/python2.7/dist-packages
Requires: jinja2, PyYAML, paramiko, pycrypto, setuptools
```

**Comment by jean-christophe-manciot** [comment]
I was able to trap the latest commit during ansible installation with ```pip install git+git://github.com/ansible/ansible.git@devel```:
```
# git log -1
commit 91449ea231757d58094eb35ec03d0af18f57c755
Author: Brian Coca <brian.coca+git@gmail.com>
Date:   Tue May 9 11:08:01 2017 -0400

    removed autogenerated file and added to ignore
```
It seems that my version is up-to-date.

**Comment by jean-christophe-manciot** [comment]
Wait, something is wrong: when I check on my locally cloned sources, I cannot find the previous commit on the devel branch:
```
git-ansible# git checkout devel
git-ansible# git log | grep 91449ea231757d58094eb35ec03d0af18f57c755
git-ansible# 
```
Something is **definitely broken with the pip builds**.

**Comment by jean-christophe-manciot** [comment]
@gundalow 
OK, I try the **git sources scripts/modules version after having uninstalled my deb and pip ansible**.
1) *2.3.1.0 RC1*
```
# ansible --version
ansible 2.3.1.0 (detached HEAD 5a3f52d988) last updated 2017/05/10 14:21:50 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/actionmystique/Ansible/git-yang-networkop/ansible-101/library']
  python version = 2.7.12+ (default, Sep 17 2016, 12:08:02) [GCC 6.2.0 20160914]
```
Running the same role/playbook on the same target leads to the same issue:
```
TASK [nxos_init : Enable NX-API] ********************************************************************************************************************************************
task path: /home/actionmystique/Ansible/Roles/roles/nxos_init/tasks/main.yml:71
<172.21.100.31> using connection plugin network_cli
fatal: [NX_OSv_Spine_31]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell", 
    "rc": 255
}
```
2) *2.4.0 (devel d3249e7875)*
```
ansible 2.4.0 (devel d3249e7875) last updated 2017/05/10 10:04:51 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/actionmystique/Ansible/git-yang-networkop/ansible-101/library']
  ansible python module location = /home/actionmystique/src/Ansible/git-ansible/lib/ansible
  executable location = /home/actionmystique/src/Ansible/git-ansible/bin/ansible
  python version = 2.7.12+ (default, Sep 17 2016, 12:08:02) [GCC 6.2.0 20160914]
```
```
TASK [nxos_init : Enable NX-API] ********************************************************************************************************************************************
task path: /home/actionmystique/Ansible/Roles/roles/nxos_init/tasks/main.yml:71
<172.21.100.31> using connection plugin network_cli
fatal: [NX_OSv_Spine_31]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell", 
    "rc": 255
}
```
Nothing has changed as far as this issue is concerned, **including the logs**, except the error message printout as @ganeshrn pointed out.
This slight difference proves that **something is wrong with**:
- **deb builds** made with ```make deb-src``` followed by ```dpkg-buildpackage --build=binary```: I thought [this issue](https://github.com/ansible/ansible/issues/22733) was solved, it turns out it is not completely reliable.
- **pip builds** installed with ```pip install git+git://github.com/ansible/ansible.git@devel```

**Comment by Qalthos** [comment]
bot_skip

**Comment by jean-christophe-manciot** [comment]
The issue is still there with:
```
ansible 2.3.2.0 (detached HEAD 022ba4fbea) last updated 2017/07/19 16:46:48 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/actionmystique/Ansible/git-yang-networkop/ansible-101/library']
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```
on the Cisco target:
```
NX-OSv9K is a demo version of the Nexus Operating System

Software
  BIOS: version 
  NXOS: version 7.0(3)I6(1)
  BIOS compile time:  
  NXOS image file is: bootflash:///nxos.7.0.3.I6.1.bin
  NXOS compile time:  5/16/2017 22:00:00 [05/17/2017 06:21:28]
```

The following role:
```
- name: Enable NX-API
  nxos_nxapi: 
      state: present 
      sandbox: "{{ sandbox }}"
      http: "{{ http.service }}"
      http_port: "{{ http.port }}"
      https: "{{ https.service }}"
      https_port: "{{ https.port }}"
      provider: "{{ connections.ssh }}"
  register: nxapi
  ignore_errors: yes
```
leads to:
```
2017-07-19 19:31:05,562 p=9922 u=root |  task path: /home/actionmystique/Ansible/Roles/roles/nxos_init/tasks/main.yml:73
2017-07-19 19:31:07,169 p=10021 u=root |  creating new control socket for host 172.21.100.18:22 as user admin
2017-07-19 19:31:07,169 p=10021 u=root |  control socket path is /root/.ansible/pc/e696d30fb8
2017-07-19 19:31:07,169 p=10021 u=root |  current working directory is /media/actionmystique/SAMSUNG5-VMs/Labs/GNS3/git-Ansible-Labs-Collection/Cisco-Nexus-9k/7.0.3.16.1.free/Ansible
2017-07-19 19:31:07,169 p=10021 u=root |  using connection plugin network_cli
2017-07-19 19:31:07,608 p=10021 u=root |  ssh connection done, setting terminal
2017-07-19 19:31:07,612 p=10021 u=root |  ssh connection has completed successfully
2017-07-19 19:31:07,612 p=10021 u=root |  connection established to 172.21.100.18 in 0:00:00.442613
2017-07-19 19:31:08,168 p=10021 u=root |  incoming request accepted on persistent socket
2017-07-19 19:31:08,168 p=10021 u=root |  socket operation is CONTEXT
2017-07-19 19:31:08,169 p=10021 u=root |  updating play_context for connection
2017-07-19 19:31:08,169 p=10021 u=root |  socket operation is EXEC
2017-07-19 19:31:08,169 p=10021 u=root |  attempting to open shell to device
2017-07-19 19:32:08,169 p=10021 u=root |  closing shell due to sigalarm
2017-07-19 19:32:08,169 p=10021 u=root |  closing shell on device
2017-07-19 19:32:08,171 p=10021 u=root |  socket operation completed with rc 255
2017-07-19 19:32:08,171 p=10021 u=root |  shutting down control socket, connection was active for 0:01:01.001547 secs
2017-07-19 19:32:08,171 p=10021 u=root |  closing connection
2017-07-19 19:32:08,171 p=10021 u=root |  closing shell on device
```
On the other hand, I am able to perform some nxos_config using NXAPI on the same NX-OSv 9k target.

**Comment by david-dumas** [comment]
I am having the same issue, ansible 2.3.1.0:
```
2017-08-09 19:01:13,301 p=5031 u=david |  using connection plugin network_cli
  5034 1502298073.36532: Loading Connection 'network_cli' from /usr/lib/python2.7/dist-packages/ansible/plugins/connection/network_cli.py
2017-08-09 19:01:13,365 p=5031 u=david |    5034 1502298073.36532: Loading Connection 'network_cli' from /usr/lib/python2.7/dist-packages/ansible/plugins/connection/network_cli.py
  5034 1502298073.36658: Loading ShellModule 'csh' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/csh.py
2017-08-09 19:01:13,366 p=5031 u=david |    5034 1502298073.36658: Loading ShellModule 'csh' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/csh.py
  5034 1502298073.36695: Loading ShellModule 'fish' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/fish.py
2017-08-09 19:01:13,367 p=5031 u=david |    5034 1502298073.36695: Loading ShellModule 'fish' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/fish.py
  5034 1502298073.36745: Loading ShellModule 'powershell' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/powershell.py
2017-08-09 19:01:13,367 p=5031 u=david |    5034 1502298073.36745: Loading ShellModule 'powershell' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/powershell.py
  5034 1502298073.36771: Loading ShellModule 'sh' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/sh.py
2017-08-09 19:01:13,367 p=5031 u=david |    5034 1502298073.36771: Loading ShellModule 'sh' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/sh.py
  5034 1502298073.36801: Loading ShellModule 'sh' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/sh.py (found_in_cache=True, class_only=False)
2017-08-09 19:01:13,368 p=5031 u=david |    5034 1502298073.36801: Loading ShellModule 'sh' from /usr/lib/python2.7/dist-packages/ansible/plugins/shell/sh.py (found_in_cache=True, class_only=False)
2017-08-09 19:01:14,110 p=5031 u=david |  ssh connection done, setting terminal
  5034 1502298074.11290: Loading TerminalModule 'nxos' from /usr/lib/python2.7/dist-packages/ansible/plugins/terminal/nxos.py
2017-08-09 19:01:14,112 p=5031 u=david |    5034 1502298074.11290: Loading TerminalModule 'nxos' from /usr/lib/python2.7/dist-packages/ansible/plugins/terminal/nxos.py
2017-08-09 19:01:14,113 p=5031 u=david |  ssh connection has completed successfully
2017-08-09 19:01:14,113 p=5031 u=david |  connection established to 192.168.1.100 in 0:00:00.811453
2017-08-09 19:01:14,300 p=5031 u=david |  incoming request accepted on persistent socket
2017-08-09 19:01:14,300 p=5031 u=david |  socket operation is CONTEXT
2017-08-09 19:01:14,301 p=5031 u=david |  updating play_context for connection
2017-08-09 19:01:14,301 p=5031 u=david |  socket operation is EXEC
2017-08-09 19:01:14,301 p=5031 u=david |  attempting to open shell to device
2017-08-09 19:01:24,301 p=5031 u=david |  closing shell due to sigalarm
2017-08-09 19:01:24,301 p=5031 u=david |  closing shell on device
2017-08-09 19:01:24,303 p=5031 u=david |  socket operation completed with rc 255
2017-08-09 19:01:24,306 p=5031 u=david |  shutting down control socket, connection was active for 0:00:11.004468 secs
2017-08-09 19:01:24,307 p=5031 u=david |  closing connection
2017-08-09 19:01:24,307 p=5031 u=david |  closing shell on device
2017-08-09 19:01:24,359 p=4957 u=david |    5030 1502298084.35957: handler run complete
2017-08-09 19:01:24,364 p=4957 u=david |    5030 1502298084.36028: attempt loop complete, returning result
2017-08-09 19:01:24,368 p=4957 u=david |    5030 1502298084.36514: _execute() done
2017-08-09 19:01:24,369 p=4957 u=david |    5030 1502298084.36878: dumping result to json
2017-08-09 19:01:24,369 p=4957 u=david |    5030 1502298084.36923: done dumping result, returning
2017-08-09 19:01:24,369 p=4957 u=david |    5030 1502298084.36959: done running TaskExecutor() for 192.168.1.100/TASK: node : IP | Route bastion public IP traffic on private interface
2017-08-09 19:01:24,372 p=4957 u=david |    5030 1502298084.37005: sending task result
2017-08-09 19:01:24,387 p=4957 u=david |    5030 1502298084.38737: done sending task result
2017-08-09 19:01:24,389 p=4957 u=david |    5030 1502298084.38909: WORKER PROCESS EXITING
2017-08-09 19:01:24,391 p=4957 u=david |    4957 1502298084.39177: marking 192.168.1.100 as failed
2017-08-09 19:01:24,392 p=4957 u=david |    4957 1502298084.39222: marking host 192.168.1.100 failed, current state: HOST STATE: block=2, task=8, rescue=0, always=0, run_state=ITERATING_TASKS, fail_state=FAILED_NONE, pending_setup=False, tasks child state? (None), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False
2017-08-09 19:01:24,393 p=4957 u=david |    4957 1502298084.39264: ^ failed state is now: HOST STATE: block=2, task=8, rescue=0, always=0, run_state=ITERATING_COMPLETE, fail_state=FAILED_TASKS, pending_setup=False, tasks child state? (None), rescue child state? (None), always child state? (None), did rescue? False, did start at task? False
2017-08-09 19:01:24,394 p=4957 u=david |    4957 1502298084.39416: getting the next task for host 192.168.1.100
2017-08-09 19:01:24,396 p=4957 u=david |    4957 1502298084.39586: host 192.168.1.100 is done iterating, returning
2017-08-09 19:01:24,397 p=4957 u=david |  fatal: [192.168.1.100]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell", 
    "rc": 255
}
```

Task:
```
- name: Add static route
  connection: local
  nxos_static_route:
    host: "{{ ansible_host }}"
    prefix: "{{ remote_ip }}/32"
    next_hop: "{{ ansible_host }}"
    route_name:: bastion```

**Comment by ganeshrn** [comment]
@jean-christophe-manciot 
```
2017-07-19 19:31:08,169 p=10021 u=root |  attempting to open shell to device
017-07-19 19:32:08,169 p=10021 u=root |  closing shell due to sigalarm
```
Based on the logs there is timeout happening after 60 sec because the response for the command executed is not received from remote device.

Just to rule out timeout issue can you please set provider timeout value to a higher value say 300.

If it still fails please run it against **latest devel branch** and share the contents of log file.

**Comment by ganeshrn** [comment]
needs_info

**Comment by hiddenicon** [comment]
@ganeshrn @jean-christophe-manciot

I'm running into this same exact issue. Only my nexus 9k's are experiencing this via CLI/SSH. All of my 5k's and 7k's are fine. 

My output from using Ansible 2.3.2.0 and using "timeout: 300" in the task:

```
19:41:59,820 paramiko.transport Authentication (password) successful!
19:41:59,821 p=14236 u=root |  ssh connection done, setting terminal
19:41:59,823 p=14236 u=root |  ssh connection has completed successfully
19:41:59,823 p=14236 u=root |  connection established to 10.60.254.161 in 0:00:00.617120
19:41:59,860 paramiko.transport Received global request "hostkeys-00@openssh.com"
19:41:59,860 paramiko.transport Rejecting "hostkeys-00@openssh.com" global request from server.
19:42:00,205 p=14236 u=root |  incoming request accepted on persistent socket
19:42:00,205 p=14236 u=root |  socket operation is CONTEXT
19:42:00,206 p=14236 u=root |  updating play_context for connection
19:42:00,206 p=14236 u=root |  socket operation is EXEC
19:42:00,206 p=14236 u=root |  attempting to open shell to device
19:42:00,207 paramiko.transport [chan 0] Max packet in: 32768 bytes
19:42:00,208 paramiko.transport [chan 0] Max packet out: 32768 bytes
19:42:00,208 paramiko.transport Secsh channel 0 opened.
19:42:00,209 paramiko.transport [chan 0] Sesch channel 0 request ok
19:42:00,211 paramiko.transport [chan 0] Sesch channel 0 request ok
19:47:00,206 p=14236 u=root |  closing shell due to sigalarm
19:47:00,207 p=14236 u=root |  closing shell on device
19:47:00,207 paramiko.transport [chan 0] EOF sent (0)
19:47:00,208 p=14236 u=root |  socket operation completed with rc 255
19:47:00,215 p=14236 u=root |  shutting down control socket, connection was active for 0:05:01.008668 secs
19:47:00,215 p=14236 u=root |  closing connection
19:47:00,215 p=14236 u=root |  closing shell on device
19:47:00,236 p=14215 u=root |  open_shell() returned 255  Traceback (most recent call last):
  File "/usr/bin/ansible-connection", line 198, in run
    (rc, stdout, stderr) = self.conn.exec_command(cmd)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 251, in exec_command
    return self.open_shell()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/__init__.py", line 54, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 106, in open_shell
    self.receive()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 144, in receive
    data = self._shell.recv(256)
AttributeError: 'NoneType' object has no attribute 'recv'
```

I updated Ansible to the latest devel, running 2.5.0. I get the same results:
```
19:51:27,923 paramiko.transport Authentication (password) successful!
19:51:27,933 p=14635 u=root |  ssh connection done, setting terminal
19:51:27,933 paramiko.transport [chan 0] Max packet in: 32768 bytes
19:51:27,935 paramiko.transport Received global request "hostkeys-00@openssh.com"
19:51:27,935 paramiko.transport Rejecting "hostkeys-00@openssh.com" global request from server.
19:51:27,975 paramiko.transport [chan 0] Max packet out: 32768 bytes
19:51:27,975 paramiko.transport Secsh channel 0 opened.
19:51:27,999 paramiko.transport [chan 0] Sesch channel 0 request ok
19:51:28,001 paramiko.transport [chan 0] Sesch channel 0 request ok
19:51:28,009 p=14635 u=root |  loaded terminal plugin for network_os nxos
19:51:28,012 p=14635 u=root |  loaded cliconf plugin for network_os nxos
19:56:27,695 p=14614 u=root |  fatal: [10.60.254.161]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"
}
2017-09-13 19:56:27,710 p=14614 u=root |  PLAY RECAP *********************************************************************
2017-09-13 19:56:27,710 p=14614 u=root |  10.60.254.161              : ok=0    changed=0    unreachable=0    failed=1   
2017-09-13 19:56:27,713 p=14614 u=root |   [WARNING]: Failure using method (v2_playbook_on_stats) in callback plugin
(<awx_display_callback.module.AWXDefaultCallbackModule object at 0x19e9810>):
'show_custom_stats'

2017-09-13 19:56:27,713 p=14614 u=root |  Callback Exception: 
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 366, in send_callback
    method(*args, **kwargs)
   File "/usr/lib/python2.7/site-packages/awx/lib/awx_display_callback/module.py", line 300, in v2_playbook_on_stats
    super(BaseCallbackModule, self).v2_playbook_on_stats(stats)
   File "/usr/lib/python2.7/site-packages/ansible/plugins/callback/default.py", line 289, in v2_playbook_on_stats
    if self._plugin_options['show_custom_stats'] and stats.custom:

2017-09-13 19:56:28,493 p=14635 u=root |  failed to create control socket for host 10.60.254.161
2017-09-13 19:56:28,494 p=14635 u=root |  Traceback (most recent call last):
  File "/usr/bin/ansible-connection", line 316, in main
    server = Server(socket_path, pc)
  File "/usr/bin/ansible-connection", line 112, in __init__
    self.connection._connect()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 139, in _connect
    self.receive()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 176, in receive
    data = self._shell.recv(256)
  File "/usr/lib/python2.7/site-packages/paramiko/channel.py", line 615, in recv
    raise socket.timeout()
timeout
```

On the same box I created a python script with paramiko and it is able to connect and run the command just fine on the 9k.
```
import paramiko
ssh = paramiko.SSHClient()
ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
ssh.connect("10.60.254.161", port=22, username="<user>", password="<pass>")
stdin, stdout, stderr = ssh.exec_command("show version")
output = stdout.readlines()
outstore = '\n'.join(output)
print output
```

**Comment by gundalow** [comment]
Hi,
Can you please test this against 2.4.0RC3?
https://groups.google.com/forum/#!topic/ansible-devel/PjBtOkwbt5I

Thanks
needs_info

**Comment by hiddenicon** [comment]
Same issue...

```
HEAD is now at 6771ade... New release v2.4.0.0-0.3.rc3

[root@awx ansible]# ansible --version
ansible 2.4.0.0 (detached HEAD 6771ade86a) last updated 2017/09/14 17:15:04 (GMT +000)
```

```
17:15:51,586 paramiko.transport Authentication (password) successful!
17:15:51,587 p=2391 u=root |  ssh connection done, setting terminal
17:15:51,587 paramiko.transport [chan 0] Max packet in: 32768 bytes
17:15:51,588 paramiko.transport Received global request "hostkeys-00@openssh.com"
17:15:51,588 paramiko.transport Rejecting "hostkeys-00@openssh.com" global request from server.
17:15:51,628 paramiko.transport [chan 0] Max packet out: 32768 bytes
17:15:51,628 paramiko.transport Secsh channel 0 opened.
17:15:51,652 paramiko.transport [chan 0] Sesch channel 0 request ok
17:15:51,655 paramiko.transport [chan 0] Sesch channel 0 request ok
17:15:51,661 p=2391 u=root |  loaded terminal plugin for network_os nxos
17:15:51,663 p=2391 u=root |  loaded cliconf plugin for network_os nxos
17:16:21,038 p=2370 u=root |  fatal: [10.60.254.161]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell"
}
2017-09-14 17:16:21,045 p=2370 u=root |  PLAY RECAP *********************************************************************
2017-09-14 17:16:21,045 p=2370 u=root |  10.60.254.161              : ok=0    changed=0    unreachable=0    failed=1   
2017-09-14 17:16:21,047 p=2370 u=root |   [WARNING]: Failure using method (v2_playbook_on_stats) in callback plugin
(<awx_display_callback.module.AWXDefaultCallbackModule object at 0x2200ad0>):
'show_custom_stats'

2017-09-14 17:16:21,048 p=2370 u=root |  Callback Exception: 
  File "/usr/lib/python2.7/site-packages/ansible/executor/task_queue_manager.py", line 366, in send_callback
    method(*args, **kwargs)
   File "/usr/lib/python2.7/site-packages/awx/lib/awx_display_callback/module.py", line 300, in v2_playbook_on_stats
    super(BaseCallbackModule, self).v2_playbook_on_stats(stats)
   File "/usr/lib/python2.7/site-packages/ansible/plugins/callback/default.py", line 289, in v2_playbook_on_stats
    if self._plugin_options['show_custom_stats'] and stats.custom:

2017-09-14 17:16:22,096 p=2391 u=root |  failed to create control socket for host 10.60.254.161
2017-09-14 17:16:22,098 p=2391 u=root |  Traceback (most recent call last):
  File "/usr/bin/ansible-connection", line 316, in main
    server = Server(socket_path, pc)
  File "/usr/bin/ansible-connection", line 112, in __init__
    self.connection._connect()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 139, in _connect
    self.receive()
  File "/usr/lib/python2.7/site-packages/ansible/plugins/connection/network_cli.py", line 176, in receive
    data = self._shell.recv(256)
  File "/usr/lib/python2.7/site-packages/paramiko/channel.py", line 667, in recv
    raise socket.timeout()
timeout
```

**Comment by jean-christophe-manciot** [comment]
@gundalow @ganeshrn  @hiddenicon 
It fails with ansible 2.4.0 RC3, but the symptom is different from [my previous post](https://github.com/ansible/ansible/issues/18974#issuecomment-316462207):, and the log is the same as hiddenicon's latest post:
```
ansible 2.4.0.0 (detached HEAD 6771ade86a) last updated 2017/09/14 18:32:30 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/actionmystique/Ansible/git-yang-networkop/ansible-101/library']
  ansible python module location = /home/actionmystique/src/Ansible/git-ansible/lib/ansible
  executable location = /home/actionmystique/src/Ansible/git-ansible/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```
With ```timeout = 300``` in /etc/ansible/ansible.cfg
With NX-OSv-9k 7.0(3)I6(1):
```
2017-09-14 19:13:09,927 p=28436 u=root |  task path: /home/actionmystique/Ansible/Roles/roles/nxos_init/tasks/main.yml:76
2017-09-14 19:13:11,738 p=28541 u=root |  creating new control socket for host 172.21.100.18:22 as user admin
2017-09-14 19:13:11,739 p=28541 u=root |  control socket path is /root/.ansible/pc/2a3fccf794
2017-09-14 19:13:11,739 p=28541 u=root |  current working directory is /media/actionmystique/SAMSUNG5-VMs/Labs/GNS3/git-Ansible-Labs-Collection/Cisco-Nexus-9k/7.0.3.16.1.free/Ansible
2017-09-14 19:13:11,739 p=28541 u=root |  using connection plugin network_cli
2017-09-14 19:13:11,819 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-09-14 19:13:11,820 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-09-14 19:13:11,839 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-09-14 19:13:11,839 paramiko.hostkeys Unable to handle key of type ssh-ed25519
2017-09-14 19:13:11,841 paramiko.transport starting thread (client mode): 0xd0aa3bd0L
2017-09-14 19:13:11,841 paramiko.transport Local version/idstring: SSH-2.0-paramiko_2.0.6
2017-09-14 19:13:12,021 paramiko.transport Remote version/idstring: SSH-2.0-OpenSSH_7.2 FIPS
2017-09-14 19:13:12,021 paramiko.transport Connected (version 2.0, client OpenSSH_7.2)
2017-09-14 19:13:12,022 paramiko.transport kex algos:[u'curve25519-sha256@libssh.org', u'ecdh-sha2-nistp256', u'ecdh-sha2-nistp384', u'diffie-hellman-group-exchange-sha256', u'diffie-hellman-group14-sha1'] server key:[u'ssh-rsa'] client encrypt:[u'aes128-ctr', u'aes192-ctr', u'aes256-ctr'] server encrypt:[u'aes128-ctr', u'aes192-ctr', u'aes256-ctr'] client mac:[u'hmac-sha1'] server mac:[u'hmac-sha1'] client compress:[u'none', u'zlib@openssh.com'] server compress:[u'none', u'zlib@openssh.com'] client lang:[u''] server lang:[u''] kex follows?False
2017-09-14 19:13:12,022 paramiko.transport Kex agreed: diffie-hellman-group14-sha1
2017-09-14 19:13:12,022 paramiko.transport Cipher agreed: aes128-ctr
2017-09-14 19:13:12,022 paramiko.transport MAC agreed: hmac-sha1
2017-09-14 19:13:12,022 paramiko.transport Compression agreed: none
2017-09-14 19:13:12,078 paramiko.transport kex engine KexGroup14 specified hash_algo <built-in function openssl_sha1>
2017-09-14 19:13:12,078 paramiko.transport Switch to new keys ...
2017-09-14 19:13:12,185 paramiko.transport userauth is OK
2017-09-14 19:13:12,199 paramiko.transport Auth banner: User Access Verification

2017-09-14 19:13:12,199 paramiko.transport Authentication (password) successful!
2017-09-14 19:13:12,213 p=28541 u=root |  ssh connection done, setting terminal
2017-09-14 19:13:12,213 paramiko.transport [chan 0] Max packet in: 32768 bytes
2017-09-14 19:13:12,214 paramiko.transport Received global request "hostkeys-00@openssh.com"
2017-09-14 19:13:12,215 paramiko.transport Rejecting "hostkeys-00@openssh.com" global request from server.
2017-09-14 19:13:12,256 paramiko.transport [chan 0] Max packet out: 32768 bytes
2017-09-14 19:13:12,257 paramiko.transport Secsh channel 0 opened.
2017-09-14 19:13:12,278 paramiko.transport [chan 0] Sesch channel 0 request ok
2017-09-14 19:13:12,283 paramiko.transport [chan 0] Sesch channel 0 request ok
2017-09-14 19:13:12,286 p=28541 u=root |  loaded terminal plugin for network_os nxos
2017-09-14 19:13:12,289 p=28541 u=root |  loaded cliconf plugin for network_os nxos
2017-09-14 19:14:11,821 p=28436 u=root |  fatal: [NX_OSv_9k_Spine_31]: FAILED! => {
```

However, I am able to manually ssh into the same target from the same host:
```
ssh admin@172.21.100.18User Access Verification
...
Cisco NX-OS Software
Copyright (c) 2002-2017, Cisco Systems, Inc. All rights reserved.
...
NX_OSv_9k_Spine_31# 
```

There is an issue with paramiko:
```
pip show paramiko
Name: paramiko
Version: 2.0.6
```

It's strange that I can manually SSH into the target, and yet paramiko used by ansible rejects ```"hostkeys-00@openssh.com" global request from server```.

**Comment by hiddenicon** [comment]
I just tried the same playbook to the 9k, but I used ios_command instead of nxos_command. It goes right in just fine and works. I'm doing a "show vesion" as the command test. 

Considering that I can SSH directly and running commands, paramiko runs just fine from a generic Python script, and it works using ios_command, it appears the issue lies somwhere in the nxos_command or nxos specific ansible modules. 

Same switch using the ios_command and still on 2.4RC3:
```
2017-09-14 17:33:03,592 paramiko.transport Authentication (password) successful!
2017-09-14 17:33:03,594 p=2750 u=root |  ssh connection done, setting terminal
2017-09-14 17:33:03,595 paramiko.transport [chan 0] Max packet in: 32768 bytes
2017-09-14 17:33:03,596 paramiko.transport Received global request "hostkeys-00@openssh.com"
2017-09-14 17:33:03,596 paramiko.transport Rejecting "hostkeys-00@openssh.com" global request from server.
2017-09-14 17:33:03,636 paramiko.transport [chan 0] Max packet out: 32768 bytes
2017-09-14 17:33:03,637 paramiko.transport Secsh channel 0 opened.
2017-09-14 17:33:03,660 paramiko.transport [chan 0] Sesch channel 0 request ok
2017-09-14 17:33:03,663 paramiko.transport [chan 0] Sesch channel 0 request ok
2017-09-14 17:33:03,669 p=2750 u=root |  loaded terminal plugin for network_os ios
2017-09-14 17:33:03,673 p=2750 u=root |  loaded cliconf plugin for network_os ios
2017-09-14 17:33:04,103 p=2750 u=root |  firing event: on_open_shell()
2017-09-14 17:33:04,526 p=2750 u=root |  firing event: on_authorize
2017-09-14 17:33:04,947 p=2750 u=root |  ssh connection has completed successfully
2017-09-14 17:33:04,947 p=2750 u=root |  connection established to 10.60.254.161 in 0:00:01.960188
2017-09-14 17:33:04,947 p=2750 u=root |  local socket is set to listening
2017-09-14 17:33:04,985 p=2750 u=root |  incoming request accepted on persistent socket
2017-09-14 17:33:04,985 p=2750 u=root |  socket operation is CONTEXT
2017-09-14 17:33:04,986 p=2750 u=root |  updating play_context for connection
2017-09-14 17:33:04,986 p=2750 u=root |  socket operation completed with rc 0
2017-09-14 17:33:04,986 p=2750 u=root |  socket operation is RUN
2017-09-14 17:33:04,986 p=2750 u=root |  socket operation completed with rc 0
2017-09-14 17:33:05,786 p=2763 u=root |  re-using existing socket for admin@10.60.254.161:22
2017-09-14 17:33:05,786 p=2750 u=root |  incoming request accepted on persistent socket
2017-09-14 17:33:05,786 p=2750 u=root |  socket operation is CONTEXT
2017-09-14 17:33:05,787 p=2750 u=root |  updating play_context for connection
2017-09-14 17:33:05,787 p=2750 u=root |  socket operation completed with rc 0
2017-09-14 17:33:05,787 p=2750 u=root |  socket operation is EXEC
2017-09-14 17:33:05,787 p=2750 u=root |  Command executed: prompt()
2017-09-14 17:33:05,787 p=2750 u=root |  socket operation completed with rc 0
2017-09-14 17:33:05,958 p=2729 u=root |  Using module file /usr/lib/python2.7/site-packages/ansible/modules/network/ios/ios_command.py
2017-09-14 17:33:06,150 p=2750 u=root |  incoming request accepted on persistent socket
2017-09-14 17:33:06,150 p=2750 u=root |  socket operation is EXEC
2017-09-14 17:33:06,150 p=2750 u=root |  Command executed: {"answer": null, "command": "show version", "prompt": null}
2017-09-14 17:33:07,137 p=2750 u=root |  socket operation completed with rc 0
2017-09-14 17:33:07,159 p=2729 u=root |  [DEPRECATION WARNING]: Param 'host' is deprecated. See the module docs for more
 information. This feature will be removed in version 2.3. Deprecation warnings
 can be disabled by setting deprecation_warnings=False in ansible.cfg.
2017-09-14 17:33:07,160 p=2729 u=root |  [DEPRECATION WARNING]: Param 'timeout' is deprecated. See the module docs for 
more information. This feature will be removed in version 2.3. Deprecation 
warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
2017-09-14 17:33:07,160 p=2729 u=root |  ok: [10.60.254.161] => {
    "changed": false, 
    "failed": false, 
    "invocation": {
        "module_args": {
            "auth_pass": null, 
            "authorize": false, 
            "commands": [
                "show version"
            ], 
            "host": "10.60.254.161", 
            "interval": 1, 
            "match": "all", 
            "password": null, 
            "port": null, 
            "provider": {
                "auth_pass": null, 
                "authorize": false, 
                "host": "10.60.254.161", 
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "port": null, 
                "ssh_keyfile": null, 
                "timeout": 30, 
                "username": "admin"
            }, 
            "retries": 10, 
            "ssh_keyfile": null, 
            "timeout": 30, 
            "username": null, 
            "wait_for": null
        }
    }, 
    "stdout": [
        "Cisco Nexus Operating System (NX-OS) Software\nTAC support: http://www.cisco.com/tac\nCopyright (C) 2002-2017, Cisco and/or its affiliates.\nAll rights reserved.\nThe copyrights to certain works contained in this software are\nowned by other third parties and used and distributed under their own\nlicenses, such as open source.  This software is provided \"as is,\" and unless\notherwise stated, there is no warranty, express or implied, including but not\nlimited to warranties of merchantability and fitness for a particular purpose.\nCertain components of this software are licensed under\nthe GNU General Public License (GPL) version 2.0 or \nGNU General Public License (GPL) version 3.0  or the GNU\nLesser General Public License (LGPL) Version 2.1 or \nLesser General Public License (LGPL) Version 2.0. \nA copy of each such license is available at\nhttp://www.opensource.org/licenses/gpl-2.0.php and\nhttp://opensource.org/licenses/gpl-3.0.html and\nhttp://www.opensource.org/licenses/lgpl-2.1.php and\nhttp://www.gnu.org/licenses/old-licenses/library.txt.\n\nSoftware\n  BIOS: version 08.32\n  NXOS: version 7.0(3)I7(1)\n  BIOS compile time:  10/18/2016\n  NXOS image file is: bootflash:///nxos.7.0.3.I7.1.bin\n  NXOS compile time:  8/31/2017 14:00:00 [08/31/2017 18:29:32]\n\n\nHardware\n  cisco Nexus9000 C9504 (4 Slot) Chassis (\"Supervisor Module\")\n  Intel(R) Xeon(R) CPU E5-2403 v2 @ 1.80GHz with 16400800 kB of memory.\n  Processor Board ID FOC21121NU1\n\n  Device name: PISC9K_CORE1_3.46.98\n  bootflash:   53298520 kB\nKernel uptime is 2 day(s), 21 hour(s), 14 minute(s), 25 second(s)\n\nLast reset at 824828 usecs after  Mon Sep 11 16:17:26 2017\n\n  Reason: Reset due to upgrade\n  System version: 7.0(3)I5(1)\n  Service: \n\nplugin\n  Core Plugin, Ethernet Plugin\n\nActive Package(s):"
    ], 
```

**Comment by hiddenicon** [comment]
@jean-christophe-manciot

If you create the following python script to use paramiko and run a command, do you get it to connect and run the command?

```
import paramiko
ssh = paramiko.SSHClient()
ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
ssh.connect("10.60.254.161", port=22, username="<user>", password="<pass>")
stdin, stdout, stderr = ssh.exec_command("show version")
output = stdout.readlines()
outstore = '\n'.join(output)
print output
```

**Comment by hiddenicon** [comment]
Looking at the code I did notice maybe a mismatch? 

In plugins/cliconf/nxos.py the method for "get_device_info" is looking at the output of show version in json format. Not all of the fields the code is looking for exists from the output on my switch. Maybe this is failing to determine what exact OS this should be connecting to?

From the outputs below, you can see that 'sys_ver_str' and 'isan_file_name' do not exist on my json output from the switch. 'chassis_id' and 'host_name' do exist however.

nxos.py:
```
    def get_device_info(self):
        device_info = {}

        device_info['network_os'] = 'nxos'
        reply = self.get(b'show version | json')
        data = json.loads(reply)

        device_info['network_os_version'] = data['sys_ver_str']
        device_info['network_os_model'] = data['chassis_id']
        device_info['network_os_hostname'] = data['host_name']
        device_info['network_os_image'] = data['isan_file_name']

        return device_info
```

Show version from my switch in json-pretty format:
```
PISC9K_CORE1_3.46.98# show version | json-pretty
{
    "header_str": "Cisco Nexus Operating System (NX-OS) Software\nTAC support: http://www.cisco.com/tac\nCopyright (C) 2002-2017, Cisco and/or its affiliates.\nAll rights reserved.\nThe copyrights t
o certain works contained in this software are\nowned by other third parties and used and distributed under their own\nlicenses, such as open source.  This software is provided \"as is,\" and unless
\notherwise stated, there is no warranty, express or implied, including but not\nlimited to warranties of merchantability and fitness for a particular purpose.\nCertain components of this software a
re licensed under\nthe GNU General Public License (GPL) version 2.0 or \nGNU General Public License (GPL) version 3.0  or the GNU\nLesser General Public License (LGPL) Version 2.1 or \nLesser Genera
l Public License (LGPL) Version 2.0. \nA copy of each such license is available at\nhttp://www.opensource.org/licenses/gpl-2.0.php and\nhttp://opensource.org/licenses/gpl-3.0.html and\nhttp://www.op
ensource.org/licenses/lgpl-2.1.php and\nhttp://www.gnu.org/licenses/old-licenses/library.txt.", 
    "bios_ver_str": "08.32", 
    "kickstart_ver_str": "7.0(3)I7(1)", 
    "bios_cmpl_time": "10/18/2016", 
    "kick_file_name": "bootflash:///nxos.7.0.3.I7.1.bin", 
    "kick_cmpl_time": "8/31/2017 14:00:00", 
    "kick_tmstmp": "08/31/2017 18:29:32", 
    "chassis_id": "Nexus9000 C9504 (4 Slot) Chassis", 
    "module_id": "Supervisor Module", 
    "cpu_name": "Intel(R) Xeon(R) CPU E5-2403 v2 @ 1.80GHz", 
    "memory": "16400800", 
    "mem_type": "kB", 
    "proc_board_id": "<SERIAL ID>", 
    "host_name": "<HOSTNAME>", 
    "bootflash_size": "53298520", 
    "kern_uptm_days": "2", 
    "kern_uptm_hrs": "22", 
    "kern_uptm_mins": "2", 
    "kern_uptm_secs": "5", 
    "rr_usecs": "824828", 
    "rr_ctime": "Mon Sep 11 20:17:26 2017", 
    "rr_reason": "Reset due to upgrade", 
    "rr_sys_ver": "7.0(3)I5(1)", 
    "rr_service": null, 
    "manufacturer": "Cisco Systems, Inc.", 
    "TABLE_package_list": {
        "ROW_package_list": {
            "package_id": null
        }
    }
}
```

**Comment by hiddenicon** [comment]
I am onto something here. 

While looking at the network_cli.py module for connection, I did some output debugs directly into the code. I noticed our plays are failing when we are hitting the `self.receive()` section of the method ` def _connect(self):`

Doing more output debugs, my play is never completing the `if prompts and not handled:` or `if self._find_prompt(window):` sections inside of the `def receive` method. 

Looking at that further, the prompt checking for that method is failing on our hostnames. Guess what, @jean-christophe-manciot and I both have "_" underscores in our hostname. I believe this is throwing the regex off when determining when the prompt appears in the receive stream. 

I changed my hostname to NOT include "_" underscores and boom it worked. Works fine. 

The question is, how is the prompt being determined when it's a prompt and can we fix that method and allow it to use "_" underscores?

Also to add, why does ios_command work but not nxos_command when there's an underscore in the hostname? Shouldn't it be the same for both modules?

**Comment by hiddenicon** [comment]
That was definitely it. I opened a new issue concerning ` "_"`  underscores in the hostname. The nxos and ios terminal files have different regexes to use for prompt checking. The ` "_" ` underscore should be added to the nxos regex. 

For the time being, I've edited my plugins/terminal/nxos.py file to be the same regexes as the ios.py file in the same directory.

Reference new issue: #30383

**Comment by jean-christophe-manciot** [comment]
Changing my underscores by '-' in my hostname has no influence on this issue.

However, I have just discovered **some undocumented API changes**: the following parameters have been silently **removed**:
- **auth_pass**
- **version**

If I remove them from my provider structure, the playbook passes:
```
< PLAY RECAP >
 ------------
NX_OSv_9k_Spine_31         : ok=9    changed=3    unreachable=0    failed=0   
```
